processor/stream/processor.go

```diff
@@ -336,41 +282,83 @@ func (p *Processor) HandleStream(ctx context.Context, ipRateLimiter *rate.Limite
 	sp, ctx := apm.StartSpan(ctx, "Stream", "Reporter")
 	defer sp.End()
 
-	for {
-
-		transformables, done := p.readBatch(ctx, ipRateLimiter, batchSize, jsonReader, res)
-		if transformables != nil {
-			err := report(ctx, publish.PendingReq{
-				Transformables: transformables,
-				Tcontext:       tctx,
-				Trace:          !sp.Dropped(),
-			})
+	var transformables []transform.Transformable
+	var done bool
+	for !done {
+		transformables, done = p.readBatch(ctx, ipRateLimiter, batchSize, sr, res)
+		if len(transformables) == 0 {
+			continue
+		}
+		// NOTE(axw) `report` takes ownership of transformables, which
+		// means we cannot reuse the slice memory. We should investigate
+		// alternative interfaces between the processor and publisher
+		// which would enable better memory reuse.
+		if err := report(ctx, publish.PendingReq{
+			Transformables: transformables,
+			Tcontext:       tctx,
+			Trace:          !sp.Dropped(),
+		}); err != nil {
+			switch err {
+			case publish.ErrChannelClosed:
+				res.Add(&Error{
+					Type:    ShuttingDownErrType,
+					Message: "server is shutting down",
+				})
+			case publish.ErrFull:
+				res.Add(&Error{
+					Type:    QueueFullErrType,
+					Message: err.Error(),
+				})
+			default:
+				res.Add(err)
+			}
+			return res
+		}
+		res.AddAccepted(len(transformables))
+	}
+	return res
+}
 
-			if err != nil {
-				switch err {
-				case publish.ErrChannelClosed:
-					res.Add(&Error{
-						Type:    ShuttingDownErrType,
-						Message: "server is shutting down",
-					})
-				case publish.ErrFull:
-					res.Add(&Error{
-						Type:    QueueFullErrType,
-						Message: err.Error(),
-					})
-				default:
-					res.Add(err)
-				}
+// streamReader wraps NDJSONStreamReader, converting errors to stream errors.
+type streamReader struct {
+	buf        *bufio.Reader
+	lineReader *decoder.LineReader
+	*decoder.NDJSONStreamReader
+}
 
-				return res
+func (sr *streamReader) Read() (map[string]interface{}, error) {
+	// TODO(axw) decode into a reused map, clearing out the
+	// map between reads. We would require that decoders copy
+	// any contents of rawModel that they wish to retain after
+	// the call, in order to safely reuse the map.
+	v, err := sr.NDJSONStreamReader.Read()
+	if err != nil {
+		if _, ok := err.(decoder.JSONDecodeError); ok {
+			return nil, &Error{
+				Type:     InvalidInputErrType,
+				Message:  err.Error(),
+				Document: string(sr.LatestLine()),
 			}
-
-			res.AddAccepted(len(transformables))
 		}
-
-		if done {
-			break
+		if err == decoder.ErrLineTooLong {
+			return nil, &Error{
+				Type:     InputTooLargeErrType,
+				Message:  "event exceeded the permitted size.",
+				Document: string(sr.LatestLine()),
+			}
 		}
 	}
-	return res
+	return v, err
+}
+
+func (sr *streamReader) Reset(r io.Reader) {
```

Comment:
I don't think this is so "low hanging"... `Reset`-like of functions introduce significant complexity because then one must not "forget" to call them, and they must be called at the right time, not messing up with order of operations, etc. I can't keep track easily of what needs to be reset, when and why.

Reply:
I've simplified things, following @simitt's suggestion. There's now just one Reset method to call, as before.